gulp-tasks/stagePr.js

```diff
@@ -0,0 +1,281 @@
+}
+
+/**
+ *
+ * @param {string} commit A full commit hash like e1d46d0aa61c14e3972f07e3cbf918424c186056
+ * @param {ReturnType<getChangedFiles>} changedFiles The files changed in this PR
+ * @param {{[key: string]: any}} build Build details returned from the Checks API
+ * @returns {string} A short message containing all information about the job
+ */
+function createAnnouncement(commit, changedFiles, build) {
+  console.log(commit, changedFiles, build);
+  const startedAt = new Date(build.started_at);
+  const completedAt = new Date(build.completed_at);
+  const duration =
+    Math.abs(startedAt.getTime() - completedAt.getTime()) / 1000 / 60;
+
+  const status = build.status;
+  const conclusion = build.conclusion;
+  if (status !== 'completed' && conclusion !== 'success') {
+    return (
+      `Staging build (${build.name}) did not succeed.` +
+      'Please ping engineering for assistance.'
+    );
+  }
+
+  const shortCommitSha = commit.substring(0, 7);
+  const baseUrl = `https://${shortCommitSha}-dot-dcc-staging.uc.r.appspot.com/`;
+
+  let announcement =
+    `Staging build (${build.name}) for commit ${commit} ` +
+    `started at ${startedAt.toString()} completed after ${duration.toFixed(
+      2
+    )} minutes.` +
+    '\n\n' +
+    `**${baseUrl}**`;
+
+  const changedPages = [];
+  if (changedFiles.static) {
+    for (const path of changedFiles.static) {
+      if (!path.includes('index.md')) {
+        continue;
+      }
+
+      const cleanPath = path.replace('site/', '').replace('index.md', '');
+      changedPages.push(`- ${baseUrl}${cleanPath}`);
+    }
+  }
+
+  if (changedPages.length) {
+    announcement +=
+      '\n\n' +
+      'The following pages likely changed with this PR:\n' +
+      changedPages.join('\n');
+  }
+
+  return announcement;
+}
+
+async function stagePr() {
+  if (!process.env.GITHUB_ACTION || !process.env.COMMIT_SHA) {
+    console.warn(
+      'This task is inteded to run on GitHub actions. Use npm run stage:personal locally instead.'
+    );
+    return;
+  }
+
+  const commit = process.env.COMMIT_SHA;
+  const changedFiles = getChangedFiles();
+
+  let triggerName = TRIGGER_NAME_STATIC_BUILD;
+  let checkName = CHECK_NAME_STATIC_BUILD;
+  let initialWait = INITIAL_WAIT_STATIC_BUILD;
+
+  // If any of the app files changed trigger a full app build instead
+  if (changedFiles.app && changedFiles.app.length) {
+    triggerName = TRIGGER_NAME_APP_BUILD;
+    checkName = CHECK_NAME_APP_BUILD;
+    initialWait = INITIAL_WAIT_APP_BUILD;
+  }
+
+  try {
+    await requestBuild(commit, triggerName);
+  } catch (e) {
+    throw Error('Failed to request staging build.');
+  }
+
+  // Wait for the webhook to create the actual
+  // build and Cloud Build propagate the status back to GitHub.
+  // This can take up to 1m30s - wait for 2m to have room to wiggle
+  // and as the build takes a while anyway
+  console.log('Waiting for Cloud Build job to start ...');
+  await wait(2 * 60 * 1000);
```

Comment:
Puh... yeah. I wanted to do this initially but didn't work towards this as the Cloud Build machines are actually beefier than the GitHub ones. That benefit gets lost with the webhook delay though.

If we can get rid of the webhook delay it might be possible the other way round: then we could think about using the Google Cloud build on GitHub as well. Not too sure about transfer times though and it would involve authenticating GitHub to the storage bucket.